src/main/scala/com/fulcrumgenomics/umi/CorrectUmis.scala

```diff
@@ -0,0 +1,229 @@
+          wrongLengthRecords += 1
+          rejectOut.foreach(w => w.addAlignment(rec))
+        }
+        else {
+          // Find matches for all the UMIs
+          val matches = sequences.map(findBestMatch(_, umiSequences))
+
+          // Update the metrics
+          matches.foreach { m =>
+            if (m.matched) {
+              val metric = umiMetrics(m.umi)
+              metric.total_matches += 1
+              m.mismatches match {
+                case 0 => metric.perfect_matches      += 1
+                case 1 => metric.one_mismatch_matches += 1
+                case 2 => metric.two_mismatch_matches += 1
+                case _ => metric.other_matches        += 1
+              }
+            }
+            else {
+              umiMetrics(unmatchedUmi).total_matches += 1
+            }
+          }
+
+          // Output the corrected read
+          if (matches.forall(_.matched)) {
+            val correctedUmi = matches.map(_.umi).mkString("-")
+            rec.setAttribute(this.umiTag, correctedUmi)
+            out.addAlignment(rec)
+          }
+          else {
+            mismatchedRecords += 1
+            rejectOut.foreach(r => r.addAlignment(rec))
+          }
+        }
+      }
+
+      progress.record(rec)
+    }
+
+    in.safelyClose()
+    out.close()
+    rejectOut.foreach(_.close())
+
+    // Finalize the metrics
+    val sortedMetrics        = umiMetrics.values.toSeq.sortBy(_.umi)
+    val totalWithUnmatched   = sortedMetrics.map(_.total_matches).sum.toDouble
+    val meanWithoutUnmatched = sortedMetrics.filter(_.umi != unmatchedUmi).map(_.total_matches).sum / (sortedMetrics.size - 1d)
+    sortedMetrics.foreach { m =>
+      m.fraction_of_matches = m.total_matches / totalWithUnmatched
+      m.representation      = m.total_matches / meanWithoutUnmatched
+    }
+
+    this.metrics.foreach(path => Metric.write(path, sortedMetrics))
+
+    // Log some summary info & warnings
+    val discarded = missingUmisRecords + wrongLengthRecords + mismatchedRecords
+    val kept      = totalRecords - discarded
+    logger.info(f"Read ${totalRecords}%,d; kept ${kept}%,d and rejected ${discarded}%,d")
+
+    if (missingUmisRecords > 0 || wrongLengthRecords > 0) {
+      logger.error("###################################################################")
+      if (missingUmisRecords > 0) logger.error(s"# ${mismatchedRecords} were missing UMI attributes in the BAM file!")
+      if (wrongLengthRecords > 0) logger.error(s"# ${wrongLengthRecords} had unexpected UMIs of differing lengths in the BAM file!")
+      logger.error("###################################################################")
+    }
+  }
+
+  /** Given a UMI sequence and a set of fixed UMIs, report the best match. */
+  private[umi] def findBestMatch(bases: String, umis: Array[String]): UmiMatch = {
+    val mismatchLimit = this.maxMismatches + minDistance
+    val mismatches    = umis.map(umi => Sequences.countMismatches(bases, umi))
```

Comment:
It seems to work nicely right now for a small-ish (10s) numbers of UMIs.  I think it's probably fine to leave it as is and tune it more if we find it slow on larger sets of UMIs later.

I always tie myself in knots with thinking about capping the number of mismatches, and I figure since UMIs are generally short (say 4-20bp) there's probably not a ton to be gained by aborting early on the counting mismatches.  It _might_ be worth trying to abort early if two sequences are found with < minMismatches, but that's a bit more complicated.